clarity-ext-scripts/clarity_ext_scripts/covid/rtpcr_analyse_script.py

```diff
@@ -1,65 +1,99 @@
+            if output.name == RT_PCR_POSITIVE_CONTROL:
+                positive_controls.append(result)
+            elif output.name == RT_PCR_NEGATIVE_CONTROL:
+                negative_controls.append(result)
+            else:
+                samples.append(result)
 
-        if not ct_analysis_service.is_valid():
-            return
+        if len(positive_controls) == 0 or len(negative_controls) == 0:
+            raise UsageError('positive and negative rtPCR controls were not found on this plate.')
 
-        # 2. Control values ok, pass this run
-        self.context.current_step.udf_rtpcr_passed = True
+        # Fetch results from service
+        result_gen = ct_analysis_service.analyze_samples(
+            positive_controls, negative_controls, samples)
 
-        # 3. Populate values
-        for _, output in self.context.all_analytes:
+        # Populate udfs
+        artifact_dict = {output.id: output for _, output in self.context.all_analytes}
+        for result in result_gen:
+            output = artifact_dict[result["id"]]
             original_sample = output.sample()
-            ct = output.udf_ct
-            covid_result = ct_analysis_service.analyze(ct, lower_bound, upper_bound)
-            original_sample.udf_map.force("rtPCR covid-19 result", covid_result)
-            original_sample.udf_map.force("rtPCR Passed", True)
+            covid_result = result[DIAGNOSIS_RESULT]
+            output.udf_map.force("rtPCR covid-19 result", covid_result)
+            output.udf_map.force("rtPCR Passed", covid_result not in FAILED_STATES)
+            original_sample.udf_map.force("rtPCR covid-19 result latest", covid_result)
+            # TODO: "rtPCR Passed" may be redundant?
+            original_sample.udf_map.force("rtPCR Passed latest", covid_result not in FAILED_STATES)
             self.context.update(original_sample)
+            self.context.update(output)
 
-    def initial_validation(self):
-        if not self._has_assay_udf():
-            raise UsageError("The udf 'Assay' must be filled in before running this script")
-        if self.assay not in BOUNDARY_DICT:
-            raise UsageError("The current assay value is not recognized: {}"
-                             .format(self.assay))
+        # Check control values
+        artifacts_that_failed = [
+            name for name in artifact_dict
+            if artifact_dict[name].udf_rtpcr_covid19_result ==
+               FAILED_ENTIRE_PLATE_BY_FAILED_EXTERNAL_CONTROL
+        ]
+        rt_pcr_passed = len(artifacts_that_failed) == 0
+        self.context.current_step.udf_map.force("rtPCR Passed", rt_pcr_passed)
+        self.context.update(self.context.current_step)
+
+    def _instantiate_service(self):
+        if self.instrument == 'RT-PCR Robot ID Covid RT-PCR':
+            service = ABI7500RTPCRAnalysisService()
+            # TODO: Assert that this is the name of the real UDF in Clarity
+            udf_ref_control_ct_value = 'HEX-CT'
```

Comment:
If you want to perhaps you could add which value to use here as a field on the AnalysisService subclasses (since they know), and then you could reuse it here instead of duplicating it.